src/neuroconv/tools/nwb_helpers/_configuration_models/_base_dataset_io.py

```diff
@@ -67,105 +84,55 @@ def _infer_dtype(dataset: Union[h5py.Dataset, zarr.Array]) -> np.dtype:
-
-    @classmethod
-    def from_neurodata_object(cls, neurodata_object: Container, field_name: str) -> "DatasetInfo":
-        location_in_file = _find_location_in_memory_nwbfile(
-            current_location=field_name, neurodata_object=neurodata_object
-        )
-        candidate_dataset = getattr(neurodata_object, field_name)
-
-        full_shape = get_data_shape(data=candidate_dataset)
-        dtype = _infer_dtype(dataset=candidate_dataset)
-
-        return cls(
-            object_id=neurodata_object.object_id,
-            object_name=neurodata_object.name,
-            location_in_file=location_in_file,
-            full_shape=full_shape,
-            dtype=dtype,
-        )
-
-
 class DatasetIOConfiguration(BaseModel, ABC):
     """A data model for configuring options about an object that will become a HDF5 or Zarr Dataset in the file."""
 
-    # TODO: When using Pydantic v2, remove
-    class Config:
-        arbitrary_types_allowed = True
+    model_config = ConfigDict(validate_assignment=True)  # Re-validate model on mutation
+
+    # Immutable fields about the dataset
+    object_id: str = Field(description="The UUID of the neurodata object containing the dataset.", frozen=True)
+    location_in_file: str = Field(
+        description=(
+            "The location of the this dataset within the in-memory NWBFile relative to the top-level root, "
+            "e.g. 'acquisition/ElectricalSeries/data'."
+        ),
+        frozen=True,
+    )
+    dataset_name: Literal["data", "timestamps"] = Field(description="The reference name of the dataset.", frozen=True)
+    dtype: InstanceOf[np.dtype] = Field(description="The data type of elements of this dataset.", frozen=True)
+    full_shape: Tuple[int, ...] = Field(description="The maximum shape of the entire dataset.", frozen=True)
 
-    dataset_info: DatasetInfo = Field(description="The immutable information about this dataset.")
-    chunk_shape: Tuple[int, ...] = Field(  # When using Pydantic v2.0, specify PositiveInt
+    # User specifiable fields
+    chunk_shape: Union[Tuple[PositiveInt, ...], None] = Field(
         description=(
             "The specified shape to use when chunking the dataset. "
             "For optimized streaming speeds, a total size of around 10 MB is recommended."
-        )
+        ),
     )
-    buffer_shape: Tuple[int, ...] = Field(
+    buffer_shape: Union[Tuple[int, ...], None] = Field(
         description=(
             "The specified shape to use when iteratively loading data into memory while writing the dataset. "
             "For optimized writing speeds and minimal RAM usage, a total size of around 1 GB is recommended."
-        )
+        ),
     )
-    # TODO: When using Pydantic v2, wrap h5py._hl.filters.FilterRefBase and numcodecs.abc.Codec with InstanceOf
-    compression_method: Union[str, h5py._hl.filters.FilterRefBase, numcodecs.abc.Codec, None] = Field(
-        default="gzip",
+    compression_method: Union[
```

Comment:
I wish we could use | instead of Union : P

Reply:
Soon

Soon...